compiler/rustc_lint/src/non_local_def.rs

```diff
@@ -232,6 +232,12 @@ impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
             ItemKind::Macro(_macro, MacroKind::Bang)
                 if cx.tcx.has_attr(item.owner_id.def_id, sym::macro_export) =>
             {
+                // this condition is very much a hack but it's the best we can do for now
+                let is_at_toplevel_doctest = self.body_depth == 2
+                    && parent_opt_item_name
+                        .map(|s| s.as_str().starts_with("_doctest"))
+                        .unwrap_or_default();
```

Comment:
What if rustdoc added an internal `#[rustc_*]` attribute to the generated doc test fns which we then look for here?

Reply:
That would be the best and that's partially what https://github.com/rust-lang/rust/pull/123974 is trying to do by making doctest regular `#[test]` fn, but there are many subtleties.

When that PR lands we could look at improving this checks. In the meantime I would like to stay with this minimal approach so it can be backported.

Reply:
Yes totally agree. Let's not add a hack to try to guess if we're inside a doctest or not please. ^^'

Reply:
I changed the PR to use the [`UNSTABLE_RUSTDOC_TEST_PATH`](https://github.com/rust-lang/rust/blob/378a43a06510f3e3a49c69c8de71745e6a884048/src/librustdoc/doctest.rs#L413C19-L413C45) env, which is already used for diagnostic code.

Reply:
Nah it's fine. Much simpler than a brand new attribute.